### ISSUE TITLE
heap-buffer-overflow in wavlike_ima_decode_block

### ISSUE THREAD

**Issue by andreafioraldi** [libsndfile/libsndfile#697]
Hi,
I found another overflow with AFL++ in wavlike_ima_decode_block (src/ima_adpcm.c:434). 

Seems very close to #687, it is always an overflow of the samples field in a loop that does blocksize iterations.

The ASan report is:

```
=================================================================
==517479==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60b000000158 at pc 0x0000005e203c bp 0x7fffffffbf10 sp 0x7fffffffbf08
WRITE of size 2 at 0x60b000000158 thread T0
    #0 0x5e203b in wavlike_ima_decode_block /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:434:26
    #1 0x5dece6 in ima_reader_init /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:244:2
    #2 0x5dda10 in wavlike_ima_init /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:120:16
    #3 0x567584 in wav_open /home/andrea/Desktop/libsndfile/src/wav.c:254:14
    #4 0x4cd5a2 in psf_open_file /home/andrea/Desktop/libsndfile/src/sndfile.c:3080:13
    #5 0x4cb92e in sf_open /home/andrea/Desktop/libsndfile/src/sndfile.c:359:9
    #6 0x4c675d in cart_dump /home/andrea/Desktop/libsndfile/programs/sndfile-info.c:479:14
    #7 0x4c4643 in main /home/andrea/Desktop/libsndfile/programs/sndfile-info.c:96:13
    #8 0x7ffff7c470b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16
    #9 0x41c44d in _start (/home/andrea/Desktop/libsndfile/programs/sndfile-info+0x41c44d)

0x60b000000158 is located 2 bytes to the right of 102-byte region [0x60b0000000f0,0x60b000000156)
allocated by thread T0 here:
    #0 0x494d02 in calloc (/home/andrea/Desktop/libsndfile/programs/sndfile-info+0x494d02)
    #1 0x5ddcc8 in ima_reader_init /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:187:16
    #2 0x5dda10 in wavlike_ima_init /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:120:16
    #3 0x567584 in wav_open /home/andrea/Desktop/libsndfile/src/wav.c:254:14
    #4 0x4cd5a2 in psf_open_file /home/andrea/Desktop/libsndfile/src/sndfile.c:3080:13
    #5 0x4cb92e in sf_open /home/andrea/Desktop/libsndfile/src/sndfile.c:359:9
    #6 0x4c675d in cart_dump /home/andrea/Desktop/libsndfile/programs/sndfile-info.c:479:14
    #7 0x4c4643 in main /home/andrea/Desktop/libsndfile/programs/sndfile-info.c:96:13
    #8 0x7ffff7c470b2 in __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:308:16

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/andrea/Desktop/libsndfile/src/ima_adpcm.c:434:26 in wavlike_ima_decode_block
Shadow bytes around the buggy address:
  0x0c167fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c167fff8000: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c167fff8010: fd fd fd fd fd fa fa fa fa fa fa fa fa fa 00 00
=>0x0c167fff8020: 00 00 00 00 00 00 00 00 00 00 06[fa]fa fa fa fa
  0x0c167fff8030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c167fff8070: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==517479==ABORTING

```

To reproduce on git master:

```
export CC='clang-10 -fsanitize=address'
export CFLAGS='-g'
./configure --disable-shared
make
./programs/sndfile-info --cart ./wavlike_ima_decode_block\ heap\ overflow
```

The testcase that triggers the bug is :
[wavlike_ima_decode_block heap overflow.tar.gz](https://github.com/libsndfile/libsndfile/files/5919065/wavlike_ima_decode_block.heap.overflow.tar.gz)

**Comment by bobsayshilol** [comment]
This appears to be the same as https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25530.

Running the test case locally on 2db6a8fd2e1e7d9d1ad309d4f4110712f7cb4e5d crashes with the above trace (as expected), and testing on branch #707 prints the following output without crashing:
```
$ ./sndfile-info --cart ../wavlike_ima_decode_block
Error : File '../wavlike' does not contain cart information.
```

**Comment by evpobr** [comment]
Fixed by #707. Thanks for report.
